diff --git a/cross-wine.conf b/cross-wine.conf
index bd5bcf3..2b4ac17 100644
--- a/cross-wine.conf
+++ b/cross-wine.conf
@@ -14,6 +14,7 @@ strip = 'strip'
 # https://github.com/mesonbuild/meson/issues/5125
 cpp_link_args = ['-mwindows']
 needs_exe_wrapper = true
+skip_sanity_check = true
 
 # For instance, this should be in the 64-bit only cross-file
 # c_args = ['-m64']
diff --git a/meson.build b/meson.build
index 67a2715..3013b60 100644
--- a/meson.build
+++ b/meson.build
@@ -150,6 +150,7 @@ if get_option('use-bitbridge')
     'boost_filesystem',
     static : true,
     dirs : [
+      '@BOOST@/lib',
       # Used by Arch based distros
       '/usr/local/lib32',
       '/usr/lib32',
@@ -163,7 +164,7 @@ if get_option('use-bitbridge')
       '/usr/lib',
     ]
   )
-  xcb_dep = winegcc.find_library('xcb')
+  xcb_dep = winegcc.find_library('xcb', dirs: [ '@XCB@/lib' ])
 
   executable(
     individual_host_name_32bit,
