--- a/src/libutil/current-process.cc	2024-08-03 20:03:24.044223025 +0900
+++ a/src/libutil/current-process.cc	2024-08-03 20:03:58.201385917 +0900
@@ -38,6 +38,11 @@
 
         auto cpuMax = readFile(cpuFile);
         auto cpuMaxParts = tokenizeString<std::vector<std::string>>(cpuMax, " \n");
+
+        if (cpuMaxParts.size() != 2) {
+            return 0;
+        }
+
         auto quota = cpuMaxParts[0];
         auto period = cpuMaxParts[1];
         if (quota != "max")
