--- a/src/modules/sni/item.cpp	2024-09-14 21:06:44.020616551 +0900
+++ b/src/modules/sni/item.cpp	2024-09-14 21:07:11.736810857 +0900
@@ -133,9 +133,6 @@
 
 void Item::setProperty(const Glib::ustring& name, Glib::VariantBase& value) {
   try {
-    spdlog::trace("Set tray item property: {}.{} = {}", id.empty() ? bus_name : id,
-                  std::string(name), get_variant<std::string>(value));
-
     if (name == "Category") {
       category = get_variant<std::string>(value);
     } else if (name == "Id") {
