--- a/pkgs/applications/emulators/wine/sources.nix	1970-01-01 09:00:01.000000000 +0900
+++ a/pkgs/applications/emulators/wine/sources.nix	1970-01-01 09:00:01.000000000 +0900
@@ -69,9 +69,9 @@
 
   unstable = fetchurl rec {
     # NOTE: Don't forget to change the hash for staging as well.
-    version = "8.20";
+    version = "8.21";
     url = "https://dl.winehq.org/wine/source/8.x/wine-${version}.tar.xz";
-    hash = "sha256-SNa3a9ZDBaICBX+8GdkiwfSWJfbRqYJZJ8ChhXJzmYI=";
+    hash = "sha256-MWJY3haVgoaLhNrut86KSnBVxzsxyGoTafzSgFMC/+U=";
     inherit (stable) patches;
 
     ## see http://wiki.winehq.org/Gecko
@@ -117,7 +117,7 @@
   staging = fetchFromGitHub rec {
     # https://github.com/wine-staging/wine-staging/releases
-    version = "8.17.1";
+    version = "8.21";
-    hash = "sha256-h36NDFYsI0y8TG41IH10IdF4QEuBkBewQ3knZ9iwDpg=";
+    hash = "sha256-s9c7hcX9qWxw39b/C1AO2wDELqwnFNwwPsKGAYYh0VM=";
     owner = "wine-staging";
     repo = "wine-staging";
     rev = "v${version}";
