--- a/pkgs/applications/emulators/wine/sources.nix	1970-01-01 09:00:01.000000000 +0900
+++ b/pkgs/applications/emulators/wine/sources.nix	1970-01-01 09:00:01.000000000 +0900
@@ -69,9 +69,9 @@
 
   unstable = fetchurl rec {
     # NOTE: Don't forget to change the hash for staging as well.
-    version = "9.0-rc1";
+    version = "9.0-rc4";
     url = "https://dl.winehq.org/wine/source/9.0/wine-${version}.tar.xz";
-    hash = "sha256-GDd3V74ffxeGzopWqET2ZXKIH/z3RcC0yjTwmmDi8zA=";
+    hash = "sha256-6/+LF2P/49HrYLEwLXimYqRhQOb3bkoMr1jT5OTJFo8=";
     inherit (stable) patches;
 
     ## see http://wiki.winehq.org/Gecko
@@ -117,7 +117,7 @@
   staging = fetchFromGitHub rec {
     # https://github.com/wine-staging/wine-staging/releases
     inherit (unstable) version;
-    hash = "sha256-jjy4r9VEJsU7FJ7RmE8+cgHTAkZVGruNUD5hzmGSB8c=";
+    hash = "sha256-NgZSYDE4TfrZjis8pxwYlsGpPgmOH994nqhbGm7lL7I=";
     owner = "wine-staging";
     repo = "wine-staging";
     rev = "v${version}";
