--- a/pkgs/applications/emulators/wine/sources.nix	1970-01-01 09:00:01.000000000 +0900
+++ a/pkgs/applications/emulators/wine/sources.nix	1970-01-01 09:00:01.000000000 +0900
@@ -69,9 +69,9 @@
 
   unstable = fetchurl rec {
     # NOTE: Don't forget to change the hash for staging as well.
-    version = "8.17";
+    version = "8.19";
     url = "https://dl.winehq.org/wine/source/8.x/wine-${version}.tar.xz";
-    hash = "sha256-8BeFvTFix05l3rE6oK3XEeN8SERUcZvt0OjCsaNGm34=";
+    hash = "sha256-klP4HIiA3cybeRSXTC/BmEhUV6R2aD+ienH+c1Iu0x0=";
     inherit (stable) patches;
 
     ## see http://wiki.winehq.org/Gecko
@@ -117,7 +117,7 @@
   staging = fetchFromGitHub rec {
     # https://github.com/wine-staging/wine-staging/releases
-    version = "8.17.1";
+    version = "8.19";
-    hash = "sha256-h36NDFYsI0y8TG41IH10IdF4QEuBkBewQ3knZ9iwDpg=";
+    hash = "sha256-TqCXYic1CWoSwrpjJ7IYCTx+GdpJ/otXNPIaBxBVKVI=";
     owner = "wine-staging";
     repo = "wine-staging";
     rev = "v${version}";
