--- a/pkgs/applications/emulators/wine/sources.nix	1970-01-01 09:00:01.000000000 +0900
+++ b/pkgs/applications/emulators/wine/sources.nix	1970-01-01 09:00:01.000000000 +0900
@@ -122,7 +122,12 @@
     repo = "wine-staging";
     rev = "v${version}";
 
-    disabledPatchsets = [ ];
+    disabledPatchsets = [
+      "eventfd_synchronization"
+      "ntdll-Junction_Points"
+      "server-File_Permissions"
+      "server-Stored_ACLs"
+    ];
   };
 
   wayland = fetchFromGitLab {
