--- a/pkgs/applications/emulators/wine/sources.nix	1970-01-01 09:00:01.000000000 +0900
+++ b/pkgs/applications/emulators/wine/sources.nix	1970-01-01 09:00:01.000000000 +0900
@@ -69,9 +69,9 @@
 
   unstable = fetchurl rec {
     # NOTE: Don't forget to change the hash for staging as well.
-    version = "9.0";
+    version = "9.1";
-    url = "https://dl.winehq.org/wine/source/9.0/wine-${version}.tar.xz";
+    url = "https://dl.winehq.org/wine/source/9.x/wine-${version}.tar.xz";
-    hash = "sha256-fP0JClOV9bdtlbtd76yKMSyN5MBwwRY7i1jaODMMpu4=";
+    hash = "sha256-AbO5G2/DXKvpOyjxkKI3upXIrHDENtkZWG3qo9olj/8=";
     inherit (stable) patches;
 
     ## see http://wiki.winehq.org/Gecko
@@ -117,7 +117,7 @@
   staging = fetchFromGitHub rec {
     # https://github.com/wine-staging/wine-staging/releases
     inherit (unstable) version;
-    hash = "sha256-lE/95OZigifreaRRCPkvA+Z0FqsBmm018jD6leSysXU=";
+    hash = "sha256-Auf3aXPYxluAC2WZvieTcN0xjWp4GzEZqywadWNYPWM=";
     owner = "wine-staging";
     repo = "wine-staging";
     rev = "v${version}";
