--- a/plugins/Common/IldaeilPlugin.cpp	2023-09-29 11:19:23.331349535 +0900
+++ b/plugins/Common/IldaeilPlugin.cpp	2023-09-29 11:24:44.401427612 +0900
@@ -70,7 +70,7 @@
         path  = water::File::getSpecialLocation(water::File::winAppData).getFullPathName() + "\\LADSPA;";
         path += water::File::getSpecialLocation(water::File::winProgramFiles).getFullPathName() + "\\LADSPA";
        #else
-        path  = getHomePath() + "/.ladspa:/usr/lib/ladspa:/usr/local/lib/ladspa";
+        path  = getHomePath() + "/.ladspa:/run/current-system/sw/lib/ladspa:/run/current-system/etc/profiles/per-user/nyarla/lib/ladspa";
        #endif
     }
 
@@ -93,7 +93,7 @@
         path  = water::File::getSpecialLocation(water::File::winAppData).getFullPathName() + "\\DSSI;";
         path += water::File::getSpecialLocation(water::File::winProgramFiles).getFullPathName() + "\\DSSI";
        #else
-        path = getHomePath() + "/.dssi:/usr/lib/dssi:/usr/local/lib/dssi";
+        path = getHomePath() + "/.dssi:/run/current-system/sw/lib/dssi:/run/current-system/etc/profiles/per-user/nyarla/lib/dssi";
        #endif
     }
 
@@ -116,7 +116,7 @@
         path  = water::File::getSpecialLocation(water::File::winAppData).getFullPathName() + "\\LV2;";
         path += water::File::getSpecialLocation(water::File::winCommonProgramFiles).getFullPathName() + "\\LV2";
        #else
-        path = getHomePath() + "/.lv2:/usr/lib/lv2:/usr/local/lib/lv2";
+        path = getHomePath() + "/.lv2:/run/current-system/sw/lib/lv2:/run/current-system/etc/profiles/per-user/nyarla/lib/lv2";
        #endif
     }
 
@@ -140,7 +140,7 @@
         path += water::File::getSpecialLocation(water::File::winProgramFiles).getFullPathName() + "\\Steinberg\\VstPlugins;";
         path += water::File::getSpecialLocation(water::File::winCommonProgramFiles).getFullPathName() + "\\VST2";
        #else
-        path = getHomePath() + "/.vst:/usr/lib/vst:/usr/local/lib/vst";
+        path = getHomePath() + "/.vst:/run/current-system/sw/lib/vst:/run/current-system/sw/lib/lxvst:/run/current-system/etc/profiles/per-user/nyarla/lib/vst:/run/current-system/etc/profiles/per-user/nyarla/lib/lxvst";
        #endif
     }
 
@@ -163,7 +163,7 @@
         path  = water::File::getSpecialLocation(water::File::winAppData).getFullPathName() + "\\VST3;";
         path += water::File::getSpecialLocation(water::File::winCommonProgramFiles).getFullPathName() + "\\VST3";
        #else
-        path = getHomePath() + "/.vst3:/usr/lib/vst3:/usr/local/lib/vst3";
+        path = getHomePath() + "/.vst3:/run/current-system/sw/lib/vst3:/run/current-system/etc/profiles/per-user/nyarla/lib/vst3";
        #endif
     }
 
@@ -186,7 +186,7 @@
         path  = water::File::getSpecialLocation(water::File::winAppData).getFullPathName() + "\\CLAP;";
         path += water::File::getSpecialLocation(water::File::winCommonProgramFiles).getFullPathName() + "\\CLAP";
        #else
-        path = getHomePath() + "/.clap:/usr/lib/clap:/usr/local/lib/clap";
+        path = getHomePath() + "/.clap:/run/current-system/sw/lib/clap:/run/current-system/etc/profiles/per-user/nyarla/lib/clap";
        #endif
     }
 
@@ -373,9 +373,9 @@
             carla_set_engine_option(fCarlaHostHandle, ENGINE_OPTION_PATH_BINARIES, 0, "/Applications/Carla.app/Contents/MacOS");
             carla_set_engine_option(fCarlaHostHandle, ENGINE_OPTION_PATH_RESOURCES, 0, "/Applications/Carla.app/Contents/MacOS/resources");
            #else
-            fDiscoveryTool = "/usr/lib/carla";
-            carla_set_engine_option(fCarlaHostHandle, ENGINE_OPTION_PATH_BINARIES, 0, "/usr/lib/carla");
-            carla_set_engine_option(fCarlaHostHandle, ENGINE_OPTION_PATH_RESOURCES, 0, "/usr/share/carla/resources");
+            fDiscoveryTool = "@carla@/lib/carla";
+            carla_set_engine_option(fCarlaHostHandle, ENGINE_OPTION_PATH_BINARIES, 0, "@carla@/lib/carla");
+            carla_set_engine_option(fCarlaHostHandle, ENGINE_OPTION_PATH_RESOURCES, 0, "@carla@/lib/carla/resources");
            #endif
         }
 
