--- a/sysmontask/cpu.py	2022-07-09 12:45:14.436138888 +0900
+++ b/sysmontask/cpu.py	2022-07-09 12:40:32.124703983 +0900
@@ -77,15 +77,15 @@
 
     # CPU Caches
     try:
-        p=popen('lscpu|grep -i -m1 "L1d"')
+        p=popen('lscpu|grep -m1 "L1d"')
         self.cpuL1LabelValue.set_text(sub("[\s]","",p.read().split(':')[1]))
         p.close()
 
-        p=popen('lscpu|grep -i -m1 "L2"')
+        p=popen('lscpu|grep -m1 "L2"')
         self.cpuL2LabelValue.set_text(sub('[\s]','',p.read().split(':')[1]))
         p.close()
 
-        p=popen('lscpu|grep -i "L3"')
+        p=popen('lscpu|grep -m1 "L3"')
         self.cpuL3LabelValue.set_text(sub('[\s]','',p.read().split(':')[1]))
         p.close()
     except:
@@ -226,4 +226,4 @@
         self.cpuUtilArray.insert(0,self.cpuUtil)
         for i in range(self.cpu_logical_cores):
             self.cpu_logical_cores_util_arrays[i].pop()
-            self.cpu_logical_cores_util_arrays[i].insert(0,temp[i])
\ ファイル末尾に改行がありません
+            self.cpu_logical_cores_util_arrays[i].insert(0,temp[i])
