--- a/CMakeLists.txt	2020-03-12 21:25:44.842237446 +0900
+++ b/CMakeLists.txt	2020-03-12 21:26:06.471434742 +0900
@@ -32,8 +32,6 @@
 
 add_subdirectory(dep)
 
-set(LOG_PATH "/var/log/violin.log" CACHE STRING "Violin log file path")
-
 option(USE_VST2SDK "Use VST2 SDK (path should be provided) instead of VeSTige (might improve compatibility)." OFF)
 
 if (USE_VST2SDK)
@@ -70,12 +68,6 @@
 
 install(DIRECTORY DESTINATION ${CMAKE_INSTALL_FULL_SYSCONFDIR}/violin)
 
-file(WRITE ${CMAKE_CURRENT_BINARY_DIR}/violin.log)
-if (NOT EXISTS ${LOG_PATH})
-   get_filename_component(LOG_PATH_DIR ${LOG_PATH} DIRECTORY)
-   install(FILES ${CMAKE_CURRENT_BINARY_DIR}/violin.log DESTINATION ${LOG_PATH_DIR} PERMISSIONS OWNER_READ OWNER_WRITE GROUP_READ GROUP_WRITE WORLD_READ WORLD_WRITE)
-endif ()
-
 # uninstall target
 if(NOT TARGET uninstall)
 	configure_file(
